### PR TITLE
feat: Added support for streaming ingestion for ADX  output plugin

### PR DESCRIPTION
**Pull request by asaharn**
# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Added support for "managed" steaming ingestion type for Azure data explorer output plugin. The data ingestion type is now configurable. The default ingestion type will be "queued" in case no type is configured. [Read more about ingestion type](https://docs.microsoft.com/en-us/azure/data-explorer/ingest-data-overview#batching-vs-streaming-ingestion)
Refactored code and added new test cases with increased code coverage to 78%

### CODE REVIEW

**Comment by srebhan** [comment]
@asaharn can you please sign the [contributor license agreement (CLA)](https://www.influxdata.com/legal/cla/)!? Otherwise we will not be able to merge your PR...

**Comment by asaharn** [comment]
@srebhan Thank you for the response.  I have signed the CLA. Do we need to check anything else?

**Comment by asaharn** [comment]
@srebhan It seems the `ci/circleci: test-go-windows` test is failing randomly. All pre checks are passing completely on local windows machine. Is there any special case we need to take case here?

**Comment by srebhan** [comment]
@asaharn restarted the flaky Windows test. If you want to look into it, help is appreciated! But that's another PR... ;-)

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 1]
```diff
@@ -75,22 +68,35 @@ func (adx *AzureDataExplorer) Connect() error {
...
-
+	adx.ingestors = make(map[string]ingest.Ingestor)
 	return nil
 }
 
+// Clean up and close the ingestor
 func (adx *AzureDataExplorer) Close() error {
+	var err error
+	for _, v := range adx.ingestors {
+		err = v.Close()
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:84]: Shouldn't you check this error? In your current code, you will only check the code of the last `adx.ingestors`...

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 2]
```diff
@@ -75,22 +68,35 @@ func (adx *AzureDataExplorer) Connect() error {
...
+	if err == nil {
+		err = err2
+	} else {
+		err = kustoerrors.GetCombinedError(err, err2)
+	}
+	if err != nil {
+		adx.Log.Warn("error closing connections")
+	} else {
+		adx.Log.Info("closed ingestor and client")
+	}
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:96]: This is hard to read, can you maybe write down the combinatorics here, comment each branch and do the right thing (e.g. return the error early)?

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 3]
```diff
@@ -103,9 +109,9 @@ func (adx *AzureDataExplorer) Write(metrics []telegraf.Metric) error {
 func (adx *AzureDataExplorer) writeTablePerMetric(metrics []telegraf.Metric) error {
 	tableMetricGroups := make(map[string][]byte)
 	// Group metrics by name and serialize them
-	for _, m := range metrics {
-		tableName := m.Name()
-		metricInBytes, err := adx.serializer.Serialize(m)
+	for _, metric := range metrics {
+		tableName := metric.Name()
+		metricInBytes, err := adx.serializer.Serialize(metric)
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:114]: Please keep the original code as the new code will likely mask the `metric` include and lead to linter issues.

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 4]
```diff
@@ -126,15 +132,14 @@ func (adx *AzureDataExplorer) writeTablePerMetric(metrics []telegraf.Metric) err
...
 	return nil
 }
 
 func (adx *AzureDataExplorer) writeSingleTable(metrics []telegraf.Metric) error {
 	//serialise each metric in metrics - store in byte[]
 	metricsArray := make([]byte, 0)
-	for _, m := range metrics {
-		metricsInBytes, err := adx.serializer.Serialize(m)
+	for _, metric := range metrics {
+		metricsInBytes, err := adx.serializer.Serialize(metric)
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:142]: See above.

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 5]
```diff
@@ -154,30 +159,32 @@ func (adx *AzureDataExplorer) writeSingleTable(metrics []telegraf.Metric) error
 func (adx *AzureDataExplorer) pushMetrics(ctx context.Context, format ingest.FileOption, tableName string, metricsArray []byte) error {
 	ingestor, err := adx.getIngestor(ctx, tableName)
 	if err != nil {
+		adx.Log.Error(err)
 		return err
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:163]: No! Whenever you return an error back to Telegraf's infrastructure it will be logged. No need for duplicate log messages here.

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 6]
```diff
@@ -154,30 +159,32 @@ func (adx *AzureDataExplorer) writeSingleTable(metrics []telegraf.Metric) error
...
 	}
-
+	length := len(metricsArray)
+	adx.Log.Debugf("Metrics array length %d for table %s", length, tableName)
 	reader := bytes.NewReader(metricsArray)
 	mapping := ingest.IngestionMappingRef(fmt.Sprintf("%s_mapping", tableName), ingest.JSON)
 	if _, err := ingestor.FromReader(ctx, reader, format, mapping); err != nil {
-		adx.Log.Errorf("sending ingestion request to Azure Data Explorer for table %q failed: %v", tableName, err)
+		adx.Log.Errorf("pushMetrics ingestion request to Azure Data Explorer for table %q failed: %v", tableName, err)
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:170]: Why is this change necessary?

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 7]
```diff
@@ -154,30 +159,32 @@ func (adx *AzureDataExplorer) writeSingleTable(metrics []telegraf.Metric) error
...
 		}
 		//create a new ingestor client for the table
-		tempIngestor, err := adx.createIngestor(adx.client, adx.Database, tableName)
+		tempIngestor, err := createIngestorByTable(adx.client, adx.Database, tableName, adx.IngestionType)
 		if err != nil {
 			return nil, fmt.Errorf("creating ingestor for %q failed: %v", tableName, err)
 		}
-		adx.ingesters[tableName] = tempIngestor
+		adx.ingestors[tableName] = tempIngestor
+		adx.Log.Infof("Ingestor for table %s created", tableName)
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:187]: Do we really need this as an info? How about making this a debug message?

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 8]
```diff
@@ -197,27 +204,31 @@ func (adx *AzureDataExplorer) createAzureDataExplorerTable(ctx context.Context,
...
-		return errors.New("Table name cannot be empty for SingleTable metrics grouping type")
+		return errors.New("table name cannot be empty for SingleTable metrics grouping type")
 	}
 	if adx.MetricsGrouping == "" {
 		adx.MetricsGrouping = tablePerMetric
 	}
+
+	if adx.IngestionType == "" {
+		adx.IngestionType = queuedIngestion
+	}
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:228]: It would be nice to check the type here. E.g.
```suggestion
	if adx.IngestionType == "" {
		adx.IngestionType = queuedIngestion
	}
	if !choice.Contains(adx.IngestionType, []string{"queued", "managed"}) {
		return fmt.Errorf("unknown ingestion type %q", adx.IngestionType)
	}
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 9]
```diff
@@ -237,8 +248,25 @@ func init() {
...
+		ingestor = mi
+	} else if strings.ToLower(ingestionType) == queuedIngestion {
+		qi, err := ingest.New(client, database, tableName, ingest.WithStaticBuffer(bufferSize, maxBuffers))
+		if err != nil {
+			return nil, err
+		}
+		ingestor = qi
+	} else {
+		err = errors.New(`ingestion_type has to be one of managed or queued`)
+	}
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:269]: Please use a switch statement here in order to allow easy extension in the future.

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks @asaharn for your contribution! I have some comments in the code, but one thing that strikes me is the massive change in the unit-tests. With this amount of change, and simultaneous changes in functionality, it is almost impossible to verify that the previous behavior did not change.

If you _really_ need to change the tests so fundamentally than I suggest the following

1. Create a PR that reworks the tests **without** changing the functionality. This way we can still verify that the tests cover the current behavior.
2. Update this PR to _not_ change _existing_ tests but only add new ones for the new behavior.

What do you think?

**Comment by asaharn** [comment]
Thank you @srebhan for you time and efforts. 
We completely agree with the point you made and really appreciate it. Here we made some high order function changes along with the added feature to make the code standardized as per our SDK, this led to addition of test cases. The changes are well tested and have gone through several level of approvals which guarantees that the existing behavior does not change, and newly added feature works as expected. 
If we make changes to segregate in separate PRs this would require a good amount of rework including the complete testing and internal PR approval processes.  
Still if it is a requirement that should be strictly followed, we can work on it.

File: plugins/outputs/azure_data_explorer/sample.conf [hunk 10]
```diff
@@ -23,3 +23,6 @@
   ## Creates tables and relevant mapping if set to true(default).
   ## Skips table and mapping creation if set to false, this is useful for running Telegraf with the lowest possible permissions i.e. table ingestor role.
   # create_tables = true
+
+ # type of ingestion managed or queued
+   # ingestion_type : ""
```
[the remark this answers is not in the document]
Comment (asaharn) [reply plugins/outputs/azure_data_explorer/sample.conf:28]: Yes. But it seems inevitable, as this is result of `make docs`  command else the pre check fails. Please take the reference of what we had earlier.

image

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go [hunk 11]
```diff
@@ -18,217 +16,290 @@ import (
...
+	require.Nil(t, adx.client)
+	require.Nil(t, adx.ingestors)
+}
 
-	errorInit := plugin.Init()
-	require.Error(t, errorInit)
-	require.Equal(t, "Endpoint configuration cannot be empty", errorInit.Error())
+type mockIngestor struct {
+	records []string
 }
```
[the remark this answers is not in the document]
Comment (asaharn) [reply plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:281]: I'm sorry I may be missing the point here. The mockIngestor is something we created newly to accommodate the test cases. 
Please correct me if the comment meant something else.

File: plugins/outputs/azure_data_explorer/sample.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/outputs/azure_data_explorer/sample.conf]: The `make docs` command only inserts _exactly_ what you have in `sample.conf`, so I guess the issue is there!?

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:296]: Well but you removed `type fakeIngestor struct` didn't you? Why? Couldn't you just modify the previously existing struct?

[the remark this answers is not in the document]
Comment (asaharn) [reply plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:296]: Thanks for the comment @srebhan. 

In ADX there is the `client` and `ingestor`. The client is responsible for the queries and management APIs of the cluster and `ingestor` is responsible for ingesting the data into the DB tables. In the previous version we have created our own fake client for the testing purpose, and this has to support 2 attributes `queries` and `internalMgmt`. In the newer GO SDKs for versions for ADX, this is replaced by the core SDKs mockClient (`kusto.NewMockClient()` ) hence the use of fakeClient was redundant. 

Since the ingestor is instantiated from the client the attributes and the behaviors also changed and hence the struct  `mockIngestor` is renamed (to be consistent with `mockClient` -> `mockIngestor` as opposed to `fakeClient` -> `fakeIngestor`). 

if it is mandatory, we can still go for rollback of the name of the struct.

**Comment by asaharn** [comment]
@srebhan  We have resolved the asked comments... 
Requesting to have a re-review...

File: plugins/outputs/azure_data_explorer/sample.conf [hunk 12]
```diff
@@ -23,3 +23,6 @@
   ## Creates tables and relevant mapping if set to true(default).
   ## Skips table and mapping creation if set to false, this is useful for running Telegraf with the lowest possible permissions i.e. table ingestor role.
   # create_tables = true
+
+  ##  Ingestion is supported through managed(streaming with fallback to batched ingestion or queued ) or queued, allowed values can be "managed" or "queued", with default to "queued"
+  # ingestion_type : ""
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/sample.conf:28]: Please follow the style found above!
```suggestion
  ##  Ingestion method to use.
  ##  Available options are
  ##    - managed  --  streaming with fallback to batched ingestion or the "queued" method below
  ##    - queued   --  queue up metrics and process sequentially
  # ingestion_type = "queued" 
```
Please check the documentation of the `queued` option (just a guess) and the default value!

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 13]
```diff
@@ -75,22 +69,38 @@ func (adx *AzureDataExplorer) Connect() error {
 		Authorizer: authorizer,
 	}
 	client, err := kusto.New(adx.Endpoint, authorization)
-
+	adx.Log.Debug("Connect : Client initialized successfully")
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:72]: Maybe just
```suggestion
	adx.Log.Debug("Client initialized successfully")
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 14]
```diff
@@ -75,22 +69,38 @@ func (adx *AzureDataExplorer) Connect() error {
...
 }
 
+// Clean up and close the ingestor
 func (adx *AzureDataExplorer) Close() error {
+	var errs []error
+	for _, v := range adx.ingestors {
+		if inerr := v.Close(); inerr != nil {
+			// accumulate errors while closing ingestors
+			errs = append(errs, inerr)
+		}
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:88]: Please use `err` for errors
```suggestion
		if err := v.Close(); err != nil {
			// accumulate errors while closing ingestors
			errs = append(errs, err)
		}
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 15]
```diff
@@ -75,22 +69,38 @@ func (adx *AzureDataExplorer) Connect() error {
...
+	}
+	if clienterr := adx.client.Close(); clienterr != nil {
+		errs = append(errs, clienterr)
+	}
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:92]:
```suggestion
	if err := adx.client.Close(); err != nil {
		errs = append(errs, err)
	}
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 16]
```diff
@@ -75,22 +69,38 @@ func (adx *AzureDataExplorer) Connect() error {
...
+		adx.Log.Warn("error closing connections")
+	} else {
+		adx.Log.Info("closed ingestor and client")
+	}
 	adx.client = nil
-	adx.ingesters = nil
-
-	return nil
+	adx.ingestors = nil
+	return comberr
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:103]: I think this is easier to read
```golang
	adx.ingestors = nil
	adx.client = nil

	if len(errs) == 0 {
		adx.Log.Info("Closed ingestors and client")
		return nil
	}

	// Combine errors into a single object and return the combined error
	return kustoerrors.GetCombinedError(errs...)
```

File: plugins/outputs/azure_data_explorer/azure_data_explorer.go [hunk 17]
```diff
@@ -237,8 +251,32 @@ func init() {
...
+		{
+			qi, err := ingest.New(client, database, tableName, ingest.WithStaticBuffer(bufferSize, maxBuffers))
+			if err != nil {
+				return nil, err
+			}
+			ingestor = qi
+		}
+	default:
+		{
+			err = fmt.Errorf(`ingestion_type has to be one of %q or %q`, managedIngestion, queuedIngestion)
```
Comment (srebhan) [inline plugins/outputs/azure_data_explorer/azure_data_explorer.go:277]: Please directly return the error here instead of risking that it get's overwritten later!

File: plugins/outputs/azure_data_explorer/azure_data_explorer_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/outputs/azure_data_explorer/azure_data_explorer_test.go:296]: No leave it as is...

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks @asaharn for the update! I have some comments in the code.

However, changing the code and the tests _at the same time_ is asking for trouble! Even though I firmly believe you about testing, it has several drawbacks

1. It is hard to follow for a reviewer and reader.
2. You might introduce issues that you are not aware of and that you (by chance) not cover during testing.
3. It makes bisecting an issue hard.

So please separate the two steps, e.g. first get in the PR modifying the test, moving from `fakeClient` to `mockIngestor`  and then in a second step introduce the streaming ingestion based on the new test framework.

**Comment by asaharn** [comment]
Hi @srebhan , Thanks much for the inputs. 

After considering the mentioned points, we have decided to move with the asked approach. 

We will be raising series of PRs as per the requirement. 

One of them Is raised here: 
https://github.com/influxdata/telegraf/pull/11697

Please let me know, if there are other changes required.
